From de865d6751ec229c8fb5bd1eb0bbeb1ad5eb7194 Mon Sep 17 00:00:00 2001
From: t <tristan75015@gmail.com>
Date: Fri, 9 Oct 2026 13:58:08 +0200
Subject: [PATCH] CMake: only the -I and -L flags of tfel-config, several
 directories

`tfel-config --includes` and `--libs` may return several directories
(for instance the headers of TDLS) and other flags: each `-I` or `-L`
flag gives one element of the list. The list of include directories is
given unquoted to CMake, and `env.sh` adds each library directory of
TFEL to `LD_LIBRARY_PATH`.
---
 cmake/env.sh.in                       |  7 ++++++-
 cmake/modules/mfront-behaviours.cmake |  4 ++--
 cmake/modules/tfel.cmake              | 27 +++++++++++++++++++--------
 3 files changed, 27 insertions(+), 11 deletions(-)

diff --git a/cmake/env.sh.in b/cmake/env.sh.in
index fa0f0ef..9f7993a 100644
--- a/cmake/env.sh.in
+++ b/cmake/env.sh.in
@@ -19,7 +19,12 @@ pathadd "@MPICXX_PATH@"
 pathadd "@CMAKE_CP_PATH@"
 export PATH
 
-ldadd "@TFEL_LIBRARY_PATH@"
+# TFEL may have several library directories (CMake list)
+IFS=';' read -ra _tfel_library_paths <<< "@TFEL_LIBRARY_PATH@"
+for _d in "${_tfel_library_paths[@]}"; do
+    ldadd "$_d"
+done
+unset _d _tfel_library_paths
 ldadd "@MFrontGenericInterface_DIR@/../../../lib"
 export LD_LIBRARY_PATH
 
diff --git a/cmake/modules/mfront-behaviours.cmake b/cmake/modules/mfront-behaviours.cmake
index f9ebca5..9818685 100644
--- a/cmake/modules/mfront-behaviours.cmake
+++ b/cmake/modules/mfront-behaviours.cmake
@@ -46,7 +46,7 @@ function(mfront_behaviours_check_library name)
   foreach(dep ${_DEPENDENCIES})
     add_mfront_dependency(${name} ${dep})
   endforeach(dep ${_DEPENDENCIES})
-  include_directories("${TFEL_INCLUDE_PATH}")
+  include_directories(${TFEL_INCLUDE_PATH})
   foreach(source ${_SOURCES})
     add_mfront_behaviour_sources(${name} ${source})
   endforeach(source ${_SOURCES})
@@ -67,7 +67,7 @@ function(mfront_behaviours_check_library name)
     add_dependencies(check ${name})
     target_include_directories(${name}
       PRIVATE "${CMAKE_CURRENT_BINARY_DIR}/${interface}/include"
-      PRIVATE "${TFEL_INCLUDE_PATH}")
+      PRIVATE ${TFEL_INCLUDE_PATH})
     if(WIN32)
       if(${CMAKE_CXX_COMPILER_ID} STREQUAL "GNU")
 	set_target_properties(${name}
diff --git a/cmake/modules/tfel.cmake b/cmake/modules/tfel.cmake
index 6ce2b86..0182d49 100644
--- a/cmake/modules/tfel.cmake
+++ b/cmake/modules/tfel.cmake
@@ -25,18 +25,29 @@ find_program(MFRONT_QUERY  mfront-query
 
 IF(TFEL_CONFIG AND MFRONT AND MFRONT_QUERY)
 
+  # `tfel-config --includes` and `tfel-config --libs` may return several
+  # directories (for instance the headers of TDLS): each `-I` or `-L` flag
+  # gives one element of the list, the other flags are ignored
   EXECUTE_PROCESS(COMMAND ${TFEL_CONFIG} "--includes"
-    OUTPUT_VARIABLE TFEL_INCLUDE_PATH
+    OUTPUT_VARIABLE _tfel_includes
     OUTPUT_STRIP_TRAILING_WHITESPACE)
-  STRING(LENGTH ${TFEL_INCLUDE_PATH}  TFEL_INCLUDE_PATH_LENGTH)
-  MATH(EXPR TFEL_INCLUDE_PATH_LENGTH "${TFEL_INCLUDE_PATH_LENGTH} - 2")
-  STRING(SUBSTRING ${TFEL_INCLUDE_PATH} 2 ${TFEL_INCLUDE_PATH_LENGTH} TFEL_INCLUDE_PATH)
+  separate_arguments(_tfel_includes UNIX_COMMAND "${_tfel_includes}")
+  set(TFEL_INCLUDE_PATH)
+  foreach(_flag IN LISTS _tfel_includes)
+    if(_flag MATCHES "^-I(.+)$")
+      list(APPEND TFEL_INCLUDE_PATH "${CMAKE_MATCH_1}")
+    endif()
+  endforeach(_flag)
   EXECUTE_PROCESS(COMMAND ${TFEL_CONFIG} "--libs"
-    OUTPUT_VARIABLE TFEL_LIBRARY_PATH
+    OUTPUT_VARIABLE _tfel_libs
     OUTPUT_STRIP_TRAILING_WHITESPACE)
-  STRING(LENGTH ${TFEL_LIBRARY_PATH}  TFEL_LIBRARY_PATH_LENGTH)
-  MATH(EXPR TFEL_LIBRARY_PATH_LENGTH "${TFEL_LIBRARY_PATH_LENGTH} - 2")
-  STRING(SUBSTRING ${TFEL_LIBRARY_PATH} 2 ${TFEL_LIBRARY_PATH_LENGTH} TFEL_LIBRARY_PATH)
+  separate_arguments(_tfel_libs UNIX_COMMAND "${_tfel_libs}")
+  set(TFEL_LIBRARY_PATH)
+  foreach(_flag IN LISTS _tfel_libs)
+    if(_flag MATCHES "^-L(.+)$")
+      list(APPEND TFEL_LIBRARY_PATH "${CMAKE_MATCH_1}")
+    endif()
+  endforeach(_flag)
 
   EXECUTE_PROCESS(COMMAND ${TFEL_CONFIG} "--cxx-standard"
     RESULT_VARIABLE TFEL_CXX_STANDARD_AVAILABLE
-- 
2.43.0

